File: tilelang/transform/add_bufstore_wrapper.py
from tvm.tirx import BufferStore, For, AttrStmt, ForKind, Var, PrimFunc, BufferLoad, Buffer, IntImm
from tvm.tirx.stmt_functor import ir_transform, post_order_visit
from tvm.tirx.transform import prim_func_pass


def AddWrapperForSingleBufStore():
    """
    Creates a TVM pass that wraps single buffer stores with parallel loops.

    This transformation adds T.Parallel wrappers around buffer stores that:
    1. Access fragment buffers only at the constant all-zero index [0, 0, ...]
    2. Are not inside existing tile operations or thread bindings

    A fragment accessed at any other index (a constant non-zero or a dynamic
    index such as a T.serial loop variable) is rejected, since a fragment is
    distributed across threads and such an access has no valid thread-ownership
    mapping in this lowering path.

    Returns:
        A prim_func_pass that applies the transformation
    """

    def pass_fn(func: PrimFunc, mod, ctx):
        # Counter for tracking nested tile operations
        tile_operation_depth = 0
        # Set of variables bound to threads
        thread_binding_vars = set()

        def get_used_variables(operation) -> set:
            """
            Collects all variables used in the given operation.

            Args:
                operation: The TIR operation to analyze

            Returns:
                Set of variables used in the operation
            """
            used_variables = set()

            def visit_variable(node):
                if isinstance(node, Var):
                    used_variables.add(node)

            post_order_visit(operation, visit_variable)
            return used_variables

        def collect_buffer_accesses(statement) -> tuple[list[Buffer], list[Buffer]]:
            """
            Categorizes buffers accessed in the statement by their scope.

            Args:
                statement: The TIR statement to analyze

            Returns:
                Tuple of (local_buffers, fragment_buffers)
            """
            accessed_buffers = set()

            def visit_buffer_access(node):
                if isinstance(node, (BufferLoad, BufferStore)):
                    accessed_buffers.add(node.buffer)

            post_order_visit(statement, visit_buffer_access)

            local_buffers = []
            fragment_buffers = []
            for buffer in accessed_buffers:
                if buffer.scope() == "local.fragment":
                    fragment_buffers.append(buffer)
                elif buffer.scope().startswith("local"):
                    local_buffers.append(buffer)
            return local_buffers, fragment_buffers

        def collect_buffer_indices(statement) -> dict[Buffer, list]:
            """
            Maps each buffer to the index vectors of every access to it.

            A buffer can be accessed more than once in a statement (e.g. a
            variable-index load and a constant-index store), so all occurrences are
            kept; recording only the last would let one slip past validation.

            Args:
                statement: The TIR statement to analyze

            Returns:
                Dictionary mapping buffers to a list of their access index vectors
            """
            buffer_to_indices = {}

            def visit_buffer_access(node):
                if isinstance(node, (BufferLoad, BufferStore)):
                    buffer_to_indices.setdefault(node.buffer, []).append(node.indices)

            post_order_visit(statement, visit_buffer_access)
            return buffer_to_indices

        def is_tile_operation_loop(loop: For) -> bool:
            """
            Determines if a For loop is a tile operation.

            Args:
                loop: The For loop to check

            Returns:
                True if the loop is a tile operation (parallel or has num_stages annotation)
            """
            return loop.kind == ForKind.PARALLEL or "num_stages" in loop.annotations

        def pre_visit(statement):
            """
            Pre-order visitor that tracks thread bindings and tile operation depth.
            """
            nonlocal tile_operation_depth

            if isinstance(statement, AttrStmt) and statement.attr_key == "thread_extent":
                thread_binding_vars.add(statement.node.var)
            elif isinstance(statement, For) and is_tile_operation_loop(statement):
                tile_operation_depth += 1

        def post_visit(statement):
            """
            Post-order visitor that applies transformations and updates counters.
            """
            nonlocal tile_operation_depth

            if isinstance(statement, For) and is_tile_operation_loop(statement):
                tile_operation_depth -= 1

            elif isinstance(statement, BufferStore):
                used_variables = get_used_variables(statement)
                thread_bound_variables = used_variables.intersection(thread_binding_vars)

                # Only transform if not inside tile operations and no thread bindings
                if tile_operation_depth == 0 and len(thread_bound_variables) == 0:
                    # Skip if no fragment buffers are accessed
                    _, fragment_buffers = collect_buffer_accesses(statement)
                    if len(fragment_buffers) == 0:
                        return statement

                    # Validate fragment buffer indices. In this fallback lowering
                    # path we only support the constant all-zero index [0, 0, ...]. A
                    # dynamic index, such as a T.serial loop variable, would
                    # address a thread-distributed fragment with no valid
                    # thread-ownership mapping and otherwise lowers to invalid
                    # code or races.
                    buffer_indices = collect_buffer_indices(statement)
                    for buffer, index_vectors in buffer_indices.items():
                        if buffer.scope() != "local.fragment":
                            continue
                        for indices in index_vectors:
                            if any(not (isinstance(index, IntImm) and index.value == 0) for index in indices):
                                index_str = ", ".join(str(index) for index in indices)
                                raise ValueError(
                                    f"Unsupported fragment access to '{buffer.name}' at index "
                                    f"[{index_str}]: only the constant all-zero index "
                                    "fragment[0, 0, ...] is supported in this fallback lowering "
                                    "path. A fragment is "
                                    "distributed across CUDA threads, so a dynamic index (e.g. a "
                                    "T.serial loop variable) or a non-zero index has no valid "
                                    "thread-ownership mapping. Use T.Parallel for elementwise "
                                    "access, or T.reduce_sum for reductions."
                                )

                    # Wrap the constant-index fragment[0, 0, ...] access with T.Parallel loop
                    return For(Var("_", "int32"), 0, 1, ForKind.PARALLEL, statement)

            return statement

        new_body = ir_transform(func.body, pre_visit, post_visit)

        return func.with_body(new_body)

    return prim_func_pass(pass_fn, opt_level=0)

File: src/transform/add_bufstore_wrapper.cc
/*!
 * \file add_bufstore_wrapper.cc
 * \brief Wrap single buffer stores with parallel loops.
 */

#include "tvm/runtime/logging.h"
#include <tvm/tir/stmt.h>
#include <tvm/tir/stmt_functor.h>
#include <tvm/tir/transform.h>

namespace tvm {
namespace tl {

using namespace tir;

using VarSet = std::unordered_set<Var>;
using BufferSet = std::unordered_set<Buffer, ObjectPtrHash, ObjectPtrEqual>;

/*!
 * \brief Helper function to calculate the intersection of two sets.
 * \tparam T The type of the elements.
 * \tparam Hash The hash function for the elements.
 * \tparam Equal The equality function for the elements.
 * \param left The first set.
 * \param right The second set.
 * \return The intersection of the two sets.
 */
template <typename T, typename Hash = std::hash<T>,
          typename Equal = std::equal_to<T>>
auto SetIntersection(const std::unordered_set<T, Hash, Equal> &left,
                     const std::unordered_set<T, Hash, Equal> &right) {
  std::unordered_set<T, Hash, Equal> result;
  // Iter the smaller set for better performance.
  const auto &smaller = left.size() <= right.size() ? left : right;
  const auto &larger = left.size() <= right.size() ? right : left;
  for (const auto &item : smaller)
    if (larger.find(item) != larger.end())
      result.insert(item);
  return result;
}

/*!
 * \brief Collect used variables in a statement.
 */
class UsedVarsCollector : StmtExprVisitor {
public:
  UsedVarsCollector() : used_variables({}) {}

  /*!
   * \brief Entry point to collect used variables.
   * \param stmt The statement to collect used variables.
   * \return The used variables.
   */
  static VarSet Collect(const Stmt &stmt) {
    UsedVarsCollector collector;
    collector.VisitStmt(stmt);
    return collector.used_variables;
  }

private:
  void VisitExpr_(const VarNode *op) final {
    used_variables.insert(ffi::GetRef<Var>(op));
  }

private:
  VarSet used_variables;
};

/*!
 * \brief Collect accessed buffers in a statement and categorizes them by their
 * scope.
 */
class AccessedBuffersCollector : public StmtExprVisitor {
public:
  AccessedBuffersCollector() = default;

  /*!
   * \brief Entry point to collect accessed buffers.
   * \param stmt The statement to collect accessed buffers.
   * \return The accessed buffers.
   */
  static std::pair<BufferSet, BufferSet> Collect(const Stmt &stmt) {
    AccessedBuffersCollector collector;
    collector.VisitStmt(stmt);
    return {collector.local_buffers, collector.fragment_buffers};
  }

private:
  /*!
   * \brief Categorize buffers by their scope.
   * \param buf The buffer to categorize.
   */
  inline void AddBuffer(const Buffer &buf) {
    if (buf.scope() == "local") {
      local_buffers.insert(buf);
    } else if (buf.scope() == "local.fragment") {
      fragment_buffers.insert(buf);
    }
  }

  void VisitExpr_(const BufferLoadNode *op) final {
    this->AddBuffer(op->buffer);
    StmtExprVisitor::VisitExpr_(op);
  }

  void VisitStmt_(const BufferStoreNode *op) final {
    this->AddBuffer(op->buffer);
    StmtExprVisitor::VisitStmt_(op);
  }

private:
  BufferSet local_buffers;
  BufferSet fragment_buffers;
};

using BufferIndicesMap = std::unordered_map<Buffer, std::vector<PrimExpr>,
                                            ObjectPtrHash, ObjectPtrEqual>;

/*!
 * \brief Collect buffer indices in a statement and return the map from buffer
 * to indices.
 */
class BufferIndicesCollector : public StmtExprVisitor {
public:
  BufferIndicesCollector() = default;

  /*!
   * \brief Entry point to collect buffer indices.
   * \param stmt The statement to collect buffer indices.
   * \return The map from buffer to indices.
   */
  static BufferIndicesMap Collect(const Stmt &stmt) {
    BufferIndicesCollector collector;
    collector.VisitStmt(stmt);
    return collector.buffer_indices;
  }

private:
  void VisitExpr_(const BufferLoadNode *op) final {
    if (buffer_indices.find(op->buffer) == buffer_indices.end())
      buffer_indices[op->buffer] = {};

    auto &vec = buffer_indices[op->buffer];
    for (const PrimExpr &idx : op->indices)
      vec.push_back(idx);

    StmtExprVisitor::VisitExpr_(op);
  }

  void VisitStmt_(const BufferStoreNode *op) final {
    if (buffer_indices.find(op->buffer) == buffer_indices.end())
      buffer_indices[op->buffer] = {};

    auto &vec = buffer_indices[op->buffer];
    for (const PrimExpr &idx : op->indices)
      vec.push_back(idx);

    StmtExprVisitor::VisitStmt_(op);
  }

private:
  BufferIndicesMap buffer_indices;
};

/*!
 * \brief Wrap single buffer stores with parallel loops.
 * \note This transformation adds T.Parallel wrappers around buffer stores that:
 *       1. Access fragment buffers with index 0
 *       2. Are not inside existing tile operations or thread bindings
 *       3. Don't access fragment buffers with non-zero indices
 */
class AddWrapperForSingleBufStore : public StmtMutator {
public:
  AddWrapperForSingleBufStore() = default;

  /*!
   * \brief Entry point to wrap single buffer stores.
   * \param func The function to wrap single buffer stores.
   * \return The function with wrapped single buffer stores.
   */
  static PrimFunc Apply(PrimFunc func) {
    AddWrapperForSingleBufStore wrapper;
    PrimFuncNode *func_node = func.CopyOnWrite();
    func_node->body = wrapper.VisitStmt(func_node->body);
    return func;
  }

private:
  /*!
   * \brief Check if the given for loop is a tile operation loop (parallel or
   * has num_stages annotation).
   * \param for_node The for loop to check.
   * \return True if the given for loop is a tile operation loop.
   */
  inline bool IsTileOperationLoop(const ForNode *for_node) {
    return for_node->kind == ForKind::kParallel ||
           for_node->annotations.find("num_stages") !=
               for_node->annotations.end();
  }

  /*!
   * \brief Check if the given attribute statement is a thread extent attribute.
   * \param attr_node The attribute statement to check.
   * \return True if the given attribute statement is a thread extent attribute.
   */
  inline bool IsThreadExtentAttr(const AttrStmtNode *attr_node) {
    return attr_node->attr_key == tir::attr::thread_extent;
  }

  Stmt VisitStmt_(const ForNode *op) final {
    bool is_tile_operation_loop = IsTileOperationLoop(op);
    if (is_tile_operation_loop)
      ++tile_operation_depth;
    Stmt new_stmt = StmtMutator::VisitStmt_(op);
    if (is_tile_operation_loop)
      --tile_operation_depth;
    return new_stmt;
  }

  Stmt VisitStmt_(const AttrStmtNode *op) final {
    if (IsThreadExtentAttr(op))
      thread_binding_vars.insert(
          ffi::GetRef<IterVar>(op->node.as<IterVarNode>())->var);
    return StmtMutator::VisitStmt_(op);
  }

  Stmt VisitStmt_(const BufferStoreNode *op) final {
    VarSet used_variables =
        UsedVarsCollector::Collect(ffi::GetRef<BufferStore>(op));
    VarSet thread_bound_variables =
        SetIntersection(used_variables, thread_binding_vars);
    // Only transform if not inside tile operations and no thread bindings
    if (tile_operation_depth == 0 && thread_bound_variables.empty()) {
      std::pair<BufferSet, BufferSet> pair =
          AccessedBuffersCollector::Collect(ffi::GetRef<BufferStore>(op));
      // Skip if no fragment buffers are accessed
      if (pair.second.empty())
        return ffi::GetRef<BufferStore>(op);
      // Validate fragment buffer indices - only index 0 is supported
      BufferIndicesMap buffer_indices =
          BufferIndicesCollector::Collect(ffi::GetRef<BufferStore>(op));
      for (const auto &[buf, indices] : buffer_indices) {
        if (buf.scope() != "local.fragment")
          continue;
        for (const auto &idx : indices) {
          bool invalid = false;
          if (const auto *imm = idx.as<IntImmNode>()) {
            invalid = imm->value != 0;
          }
          LOG_IF(FATAL, invalid)
              << "Fragment buffer access with non-zero index " << idx
              << " is not supported. Only fragment[0] access is allowed.";
        }
      }
      // Wrap fragment[0] access with T.Parallel loop
      return For(Var{"_"}, 0, 1, ForKind::kParallel,
                 ffi::GetRef<BufferStore>(op));
    }
    return ffi::GetRef<BufferStore>(op);
  }

private:
  /*! \brief Counter for tracking nested tile operations. */
  size_t tile_operation_depth = 0;
  /*! \brief Set of thread binding variables. */
  VarSet thread_binding_vars;
};

tvm::transform::Pass AddWrapperForSingleBufStorePass() {
  using namespace tir::transform;
  auto pass_func = [](PrimFunc f, const IRModule &, PassContext) {
    return AddWrapperForSingleBufStore::Apply(std::move(f));
  };
  return CreatePrimFuncPass(pass_func, 0, "tl.AddWrapperForSingleBufStore", {});
}

TVM_FFI_STATIC_INIT_BLOCK() {
  namespace refl = tvm::ffi::reflection;
  refl::GlobalDef().def("tl.transform.AddWrapperForSingleBufStore",
                        AddWrapperForSingleBufStorePass);
}

} // namespace tl
} // namespace tvm
